Transition AcceleratedEffectValues to using absolute lengths

weinig opened the pull request:
#### 54df9858341fa03c1f27cd119b8ce4017b02217f
<pre>
Transition AcceleratedEffectValues to using absolute lengths
<a href="https://bugs.webkit.org/show_bug.cgi?id=300042">https://bugs.webkit.org/show_bug.cgi?id=300042</a>

Reviewed by Antoine Quint and Antti Koivisto.

Started adding property specific types for AcceleratedEffect values,
initially focusing on removing the remaining uses of WebCore::Length
and WebCore::LengthPoint in AcceleratedEffectValues.

Added layering violation comments to indicate further work needed
to resolve the layering.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/LengthFunctions.cpp:
* Source/WebCore/platform/LengthFunctions.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/animation/AnimationUtilities.h:
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetAnchor.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetAnchor.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetDistance.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetDistance.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetPosition.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetPosition.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetRotate.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetRotate.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOpacity.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOpacity.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformBox.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformOrigin.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformOrigin.h: Added.
* Source/WebCore/rendering/MotionPath.cpp:
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/color/StyleOpacity.h:
* Source/WebCore/style/values/motion/StyleOffsetAnchor.cpp:
* Source/WebCore/style/values/motion/StyleOffsetAnchor.h:
* Source/WebCore/style/values/motion/StyleOffsetDistance.h:
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/motion/StyleOffsetPosition.h:
* Source/WebCore/style/values/motion/StyleOffsetRotate.cpp:
* Source/WebCore/style/values/motion/StyleOffsetRotate.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:

Canonical link: <a href="https://commits.webkit.org/300956@main">https://commits.webkit.org/300956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48c3f48b7f5159c1cbb7d611456f8ea7c01061f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124418 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34833 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131260 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126295 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44848 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94654 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35749 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75231 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34684 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29447 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74741 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105501 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133925 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51319 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39150 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103132 "Failed to checkout and rebase branch from PR 51705") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107502 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102920 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26203 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48288 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26543 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48266 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51185 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56971 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50606 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52280 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/platform/animation/values/AcceleratedEffectOffsetAnchor.h:
```diff
@@ -0,0 +1,50 @@
+
+// MARK: - Blending
+
+auto canBlend(const AcceleratedEffectOffsetAnchor&, const AcceleratedEffectOffsetAnchor&) -> bool;
```
I notice you seem to use the `auto foo() -> T;` style over `T foo();`. As I understand it, these are functionally equivalent except in the case where namespace or class qualifiers for types outside of the current scope where the former allows less typing than the latter. This is not material to reviewing this patch, but I wonder if we should discuss this more widely with respect to our coding style.

graouts replied:
Turns out this is covered by our [coding style guide](https://webkit.org/code-style-guidelines/#function-return-arrow) and we should be using the `T foo();` style here.

weinig replied:
Fair enough. I have definitely been breaking this rule a lot lately, but if I want to do that, I should clearly start a discussion first.

graouts commented on Source/WebCore/platform/animation/values/AcceleratedEffectOffsetDistance.h:
```diff
@@ -0,0 +1,47 @@
+
+struct BlendingContext;
+
+struct AcceleratedEffectOffsetDistance {
+    float value { 0 };
+
+    constexpr bool operator==(const AcceleratedEffectOffsetDistance&) const = default;
+};
```
Do we need this type at all if it's just a wrapper for a `float`?

weinig replied:
I think there a few reasons to keep these wrapper structs for cases like this:

1. Documents what the value is for explicitly.
2. Disallows using the value in the wrong circumstances. For instance, can't use an `AcceleratedEffectOffsetDistance` accidentally as a `AcceleratedEffectOpacity`, even though they are both `float` underneath it all.
3. Allow easy updating if the underlying type needs to change in the future.

There is also very little downside, since its zero cost at compile time, and the headers contain almost no includes (though I do see I included FloatSize unnecessarily in a few which I will fix).

graouts replied:
Works for me!

graouts commented on Source/WebCore/platform/animation/values/AcceleratedEffectOpacity.h:
```diff
@@ -0,0 +1,48 @@
+
+struct BlendingContext;
+
+struct AcceleratedEffectOpacity {
+    float value { 1 };
+
+    constexpr bool operator==(const AcceleratedEffectOpacity&) const = default;
+};
```
I have the same comment about this class as I do about the `AcceleratedEffectOffsetDistance` class, perhaps we can do away with it?

graouts commented on Source/WebCore/platform/animation/values/AcceleratedEffectTransformOrigin.h:
```diff
@@ -0,0 +1,50 @@
+
+struct BlendingContext;
+
+struct AcceleratedEffectTransformOrigin {
+    FloatPoint value { };
+
+    constexpr bool operator==(const AcceleratedEffectTransformOrigin&) const = default;
+};
```
Same comment here as with `AcceleratedEffectOffsetDistance` and `AcceleratedEffectOpacity`. Could we use a `FloatPoint` directly where `AcceleratedEffectTransformOrigin` is used?

graouts commented on Source/WebCore/platform/animation/AcceleratedEffect.cpp:
```diff
@@ -361,7 +361,7 @@ void AcceleratedEffect::apply(WebAnimationTime currentTime, AcceleratedEffectVal
     }();
 
-    // FIXME: when we add threaded animaiton support support for scroll-driven animations,
+    // FIXME: when we add threaded animation support support for scroll-driven animations,
```
Nice.

graouts commented on Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
```diff
@@ -46,42 +45,27 @@ namespace WebCore {
-    if (RefPtr srcOffsetPath = offsetPath)
-        clonedOffsetPath = srcOffsetPath->clone();
-
+    RefPtr clonedTranslate = translate ? RefPtr { translate->clone() } : nullptr;
+    RefPtr clonedScale = scale ? RefPtr { scale->clone() } : nullptr;
+    RefPtr clonedRotate = rotate ? RefPtr { rotate->clone() } : nullptr;
+    RefPtr clonedOffsetPath = offsetPath ? RefPtr { offsetPath->clone() } : nullptr;
```
Very nice.

graouts reviewed (approved):
Aside from small comments, looks good to me for the animation-related parts. Would be nice for someone else to review the style parts, perhaps @anttijk?

weinig committed:
Transition AcceleratedEffectValues to using absolute lengths
https://bugs.webkit.org/show_bug.cgi?id=300042

Reviewed by Antoine Quint and Antti Koivisto.

Started adding property specific types for AcceleratedEffect values,
initially focusing on removing the remaining uses of WebCore::Length
and WebCore::LengthPoint in AcceleratedEffectValues.

Added layering violation comments to indicate further work needed
to resolve the layering.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/LengthFunctions.cpp:
* Source/WebCore/platform/LengthFunctions.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/animation/AnimationUtilities.h:
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetAnchor.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetAnchor.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetDistance.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetDistance.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetPosition.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetPosition.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetRotate.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOffsetRotate.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOpacity.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectOpacity.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformBox.h: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformOrigin.cpp: Added.
* Source/WebCore/platform/animation/values/AcceleratedEffectTransformOrigin.h: Added.
* Source/WebCore/rendering/MotionPath.cpp:
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/color/StyleOpacity.h:
* Source/WebCore/style/values/motion/StyleOffsetAnchor.cpp:
* Source/WebCore/style/values/motion/StyleOffsetAnchor.h:
* Source/WebCore/style/values/motion/StyleOffsetDistance.h:
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/motion/StyleOffsetPosition.h:
* Source/WebCore/style/values/motion/StyleOffsetRotate.cpp:
* Source/WebCore/style/values/motion/StyleOffsetRotate.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:

Canonical link: https://commits.webkit.org/300956@main
